[AMDGPU] Allocate i1 argument to SGPRs

jwanggit86 opened the pull request:
Currently i1 arguments are passed as 32-bit VGPRs. It would make more sense to make use of SGPRs and pass these values as a wavesize bool mask.

jwanggit86 committed:
[AMDGPU] Allocate i1 argument to SGPRs

Currently i1 arguments are passed as 32-bit VGPRs. It would make more
sense to make use of SGPRs and pass these values as a wavesize bool mask.

jwanggit86 committed:
Fix format.

jwanggit86 commented:
Some explanations. First the calling convention is changed such that i1 arg is allocated to SGPR_32. Then 2 problems have to be fixed to prevent compile errors: (1) on the callee side, handle SGPR_32 to SGPR_64 copies by copying the SGPR_32 to the lower 32b of the SGPR_64, and set the higher 32b to 0 (2) on the caller side, handle SGPR_64 to SGPR_32 copies by ignoring the higher 32b of the SGPR_64.

I'm not sure if there are any potential problems with all the code changes and can't help thinking there might be a better solution. So, I'd like to have some comments before going any further. Thanks!

perlfu commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -872,19 +882,33 @@ void SIInstrInfo::copyPhysReg(MachineBasicBlock &MBB,
       if (AMDGPU::SReg_64RegClass.contains(SrcReg)) {
         BuildMI(MBB, MI, DL, get(AMDGPU::S_MOV_B64), AMDGPU::VCC)
-          .addReg(SrcReg, getKillRegState(KillSrc));
+            .addReg(SrcReg, getKillRegState(KillSrc));
```
This whitespace change should probably not be in this diff.

perlfu commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -872,19 +882,33 @@ void SIInstrInfo::copyPhysReg(MachineBasicBlock &MBB,
         BuildMI(MBB, MI, DL, get(AMDGPU::V_CMP_NE_U32_e32))
-          .addImm(0)
-          .addReg(SrcReg, getKillRegState(KillSrc));
+            .addImm(0)
```
As above.

perlfu commented on llvm/test/CodeGen/AMDGPU/z_callee.ll:
```diff
@@ -0,0 +1,33 @@
+
+define void @void_func_i1(i1 %arg0) #0 {
+; For CIGFX89, the i1 arg is passed in s4, but the v_cndmask insn uses s[4:5].
+; Therefore, the "s_mov_b32 s5, 0" is generated.
```
This doesn't make sense, the i1 argument represents 64 lanes of bitmask so must span s[4:5] (for GFX89).
Zeroing s5 loses half the data?

perlfu reviewed (commented):
Needs test for i1 return values?

Patch doesn't seem wave size sensitive as it just attempts to fix up COPYs from s32 to s64 and vice versa -- but based on the test these COPYs seem to represent a deeper problem that is not being addressed.

arsenm commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -852,12 +852,22 @@ void SIInstrInfo::copyPhysReg(MachineBasicBlock &MBB,
     }
 
     if (!AMDGPU::SReg_32RegClass.contains(SrcReg)) {
+      // When calling convention allocates SGPR for i1 argument, we may
+      // have a SRPR_64 to SReg_32 copy for an outgoing i1 argument. Adjust
+      // the copy to avoid illegal copy.
+      if (AMDGPU::SGPR_64RegClass.contains(SrcReg)) {
+        auto sub0 = RI.getSubReg(SrcReg, AMDGPU::sub0);
+        if (sub0 != DestReg)
+          BuildMI(MBB, MI, DL, get(AMDGPU::S_MOV_B32), DestReg).addReg(sub0);
+        return;
+      }
+
```
This is just broken, you shouldn't have to touch any MIR code. Such a copy should fail the verifier

arsenm commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -872,19 +882,33 @@ void SIInstrInfo::copyPhysReg(MachineBasicBlock &MBB,
+      // When an i1 argument is allocated to an SGPR_32, we may have a COPY
+      // from SGPR_32 to SReg_64. The following handles this case to avoid
+      // an illegal copy.
+      if (AMDGPU::SGPR_32RegClass.contains(SrcReg)) {
```
Ditto, this is just broken. You shouldn't require any MIR fixups, only ABI lowering in SelectionDAG/GlobalISel

arsenm commented on llvm/lib/Target/AMDGPU/SILowerI1Copies.cpp:
```diff
@@ -513,6 +513,12 @@ bool SILowerI1Copies::lowerCopiesFromI1() {
+      // When the calling convention allocates i1 argument to SGPR,
+      // we may have a COPY with dst being an SGPR_32. This should
+      // not be lowered into V_CNDMASK_B32.
+      if (AMDGPU::SGPR_32RegClass.contains(DstReg))
```
Ditto, this is wrong

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallingConv.td:
```diff
@@ -185,9 +185,12 @@ def CSR_AMDGPU_NoRegs : CalleeSavedRegs<(add)>;
   CCIfType<[i8, i16], CCIfExtend<CCPromoteToType<i32>>>,
 
+  CCIfType<[i1] , CCAssignToReg<
+    !foreach(i, !range(0, 30), !cast<Register>("SGPR"#i))  // SGPR0-29
```
This isn't really right, the register used differs based on the wavesize. i.e. wave32 should start at s0, and wave64 would use s[0:1] (ignoring the registers used for the SRD of course).

This may require some hacking around in the lowering code. We may also need to define separate wave size calling CallingConv variants

jwanggit86 replied:
See latest solution.

tsymalla commented on llvm/lib/Target/AMDGPU/SIInstrInfo.cpp:
```diff
@@ -852,12 +852,22 @@ void SIInstrInfo::copyPhysReg(MachineBasicBlock &MBB,
 
     if (!AMDGPU::SReg_32RegClass.contains(SrcReg)) {
+      // When calling convention allocates SGPR for i1 argument, we may
+      // have a SRPR_64 to SReg_32 copy for an outgoing i1 argument. Adjust
```
Typo: SRPR_64

jwanggit86 replied:
This change has been reverted in the latest solution.

jwanggit86 committed:
Creating a custom calling conv function for i1.

jwanggit86 committed:
Fix formatting.

jwanggit86 commented:
The latest solution is quite different from the previous one. For calling convention, a custom function is created for i1 arg/return, which allocates i1 to either a 32b SGPR for wavesize32 or a 64b SGPR for wavesize64.

Two additional changes are made: (1) in LowerFormalArgument(), previously it's assumed the all registers are from a 32b regclass. Now it's possible that it is from a 64b regclass. (2) In lowerCopiesToI1(), for i1 return value, previously it is assumed it's only from a 32b register. Now it can be from a 64b register.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -29,6 +29,39 @@
+                                CCValAssign::LocInfo LocInfo,
+                                ISD::ArgFlagsTy ArgFlags, CCState &State) {
+  static bool IsWave64 = static_cast<const GCNSubtarget &>(
+                             State.getMachineFunction().getSubtarget())
```
getSubtarget<GCNSubtarget.() is cleaner than using your own static_cast

jwanggit86 replied:
Done.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -29,6 +29,39 @@
+                             State.getMachineFunction().getSubtarget())
+                             .isWave64();
+
+  static const MCPhysReg I1RegList1[] = {
```
Name these something descriptive, SGPRArgsWave32/SGPRArgsWave64?

jwanggit86 replied:
Done.

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -2821,8 +2821,13 @@ SDValue SITargetLowering::LowerFormalArguments(
-    else
-      llvm_unreachable("Unexpected register class in LowerFormalArguments!");
+    else {
+      if (VT == MVT::i1 && Subtarget->isWave64())
```
Just use getBoolRC?

jwanggit86 replied:
Done.

arsenm commented on llvm/lib/Target/AMDGPU/SILowerI1Copies.cpp:
```diff
@@ -715,6 +715,14 @@ bool SILowerI1Copies::lowerCopiesToI1() {
       assert(!MI.getOperand(1).getSubReg());
 
       if (!SrcReg.isVirtual() || (!isLaneMaskReg(SrcReg) && !isVreg1(SrcReg))) {
+        if (!SrcReg.isVirtual() &&
+            TII->getRegisterInfo().getRegSizeInBits(SrcReg, *MRI) == 64) {
+          // When calling convention allocates SGPR for i1, for GPUs with
+          // wavefront size 64, i1 return value is put in 64b SGPR.
+          assert(ST->isWave64());
+          continue;
+        }
+
```
You don't know this is the context at this point. I don't think you should need to touch this. I assume you encountered a problematic copy here, what exactly does it look like?

jwanggit86 replied:
Immediate after these new lines is an assert():
`assert(TII->getRegisterInfo().getRegSizeInBits(SrcReg, *MRI) == 32);`
which makes sure SrcReg is 32b before lowering the copy to `V_CMP_NE_U32_e64`.

Because, with the callingconv change, the i1 return value is in an SGPR_64 (for wave64 GPUs), these lines were inserted to avoid triggering the assert.

arsenm replied:
But contextually these shouldn't be getting processed. What is the MIR at the point of the assert?

jwanggit86 replied:
The IR instruction callseq_end is always divergent. As a result, the CopyFromReg to copy the return value is also divergent. When the return type is i1, in the machine instruction, the dest reg would be vreg_1. In the above code in SILoweringI1Copies, we avoid lowing this COPY to avoid triggering the assert (code following assert assumes the src is a 32b phys reg).

arsenm replied:
Can you post the actual MIR you are getting? You really should not touch any code in SILowerI1Copies, you do not have the correct context here to know if you're doing something sensible.

callseq_end isn't associated with a value, it's just the SP modification and provides ordering. It's disconnected from the passed register value. In this context, the raw register copy is uniform, but the logical value is divergent. Coming of of the DAG for an incoming value, I would expect you to end up with something like:

```
  %0:sreg_64_exec = COPY $sgpr0_sgpr1 # This ABI copy should exist as-is
  %1:vreg_1 = COPY %0 # SILowerI1Copies is supposed to deal with this 
```

For the outgoing side, I would expect:

```
  %2:vreg_1 = ...
  %3:sreg_64_exec = COPY %2 
  $sgpr0_sgpr1 = COPY %3 # Untouchable ABI copy
```

jwanggit86 replied:
The machine code when reaching SILowerI1Copies is as follows:
```
# Machine code for function test_call_external_i1_func_void: IsSSA, TracksLiveness

bb.0 (%ir-block.0):
  ADJCALLSTACKUP 0, 0, implicit-def dead $scc
  %0:sreg_64 = SI_PC_ADD_REL_OFFSET target-flags(amdgpu-gotprel32-lo) @i1_func_void, target-flags(amdgpu-gotprel32-hi) @i1_func_void, implicit-def dead $scc
  %1:sreg_64_xexec = S_LOAD_DWORDX2_IMM killed %0:sreg_64, 0, 0 :: (dereferenceable invariant load (s64) from got, addrspace 4)
  %2:sgpr_128 = COPY $sgpr0_sgpr1_sgpr2_sgpr3
  $sgpr0_sgpr1_sgpr2_sgpr3 = COPY %2:sgpr_128
  SI_CALL_ISEL killed %1:sreg_64_xexec, @i1_func_void, <regmask $sgpr_null $sgpr_null_hi $src_private_base $src_private_base_hi $src_private_base_lo $src_private_limit $src_private_limit_hi $src_private_limit_lo $src_shared_base $src_shared_base_hi $src_shared_base_lo $src_shared_limit $src_shared_limit_hi $src_shared_limit_lo $sgpr30 $sgpr31 $sgpr32 $sgpr33 $sgpr34 $sgpr35 $sgpr36 $sgpr37 $sgpr38 $sgpr39 $sgpr40 $sgpr41 $sgpr42 $sgpr43 $sgpr44 $sgpr45 $sgpr46 $sgpr47 $sgpr48 and 1194 more...>, implicit $sgpr0_sgpr1_sgpr2_sgpr3, implicit-def $sgpr0_sgpr1
  ADJCALLSTACKDOWN 0, 0, implicit-def dead $scc
  %3:vreg_1 = COPY $sgpr0_sgpr1
  %5:sreg_64_xexec = COPY %3:vreg_1
  %4:vgpr_32 = V_CNDMASK_B32_e64 0, 0, 0, 1, %5:sreg_64_xexec, implicit $exec
  %6:sreg_64 = IMPLICIT_DEF
  %7:vreg_64 = COPY %6:sreg_64
  GLOBAL_STORE_BYTE killed %7:vreg_64, killed %4:vgpr_32, 0, 0, implicit $exec :: (volatile store (s8) into `ptr addrspace(1) undef`, addrspace 1)
  SI_RETURN
```
In the COPY of $sgpr0_sgpr1 to %3, the dest reg is vreg_1, and the source is a 64b phys reg. The above code change is to skip this instruction in order to avoid triggering the assert which ensures the src is 32b.

jwanggit86 replied:
```
      MRI->setRegClass(DstReg, IsWave32 ? &AMDGPU::SReg_32RegClass
                                        : &AMDGPU::SReg_64RegClass);
      ...      
      if (!SrcReg.isVirtual() || (!isLaneMaskReg(SrcReg) && !isVreg1(SrcReg))) {
        if (!SrcReg.isVirtual() &&
            TII->getRegisterInfo().getRegSizeInBits(SrcReg, *MRI) == 64) {
          // When calling convention allocates SGPR for i1, for GPUs with
          // wavefront size 64, i1 return value is put in 64b SGPR.
          assert(ST->isWave64());
          continue;
        }

        assert(TII->getRegisterInfo().getRegSizeInBits(SrcReg, *MRI) == 32);
        ...
```
Actually before the IF on line 717 there is a call of `setRegClass` which changes the dest reg
of %3 from vreg_1 to sreg_64. So when line 718 is reached, the insn is already updated to 
`"%3: sreg_64 = COPY $sgpr0_sgpr1"`. The lines added by this patch (718-725) is simply to avoid
triggering the assert on line 726 which assumes that the src phys reg is 32b. For wave64 GPUs, with the calling
conv change introduced by this patch, that assumption is no longer valid.

jwanggit86 replied:
@arsenm Is the code in SILowerI1Copies ok, or do you still have questions?

arsenm replied:
I still think that both:

1. The code already in SILowerI1Copies is nonsense which should be fixed
2. You shouldn't need to touch it

I think the code that's reaching here isn't what you want. You want to have the original copy be the copy to the SReg_64_xexec/SReg_32_xexec, and a copy from there to VReg_1. You have the opposite copy pair here

jwanggit86 replied:
> 1. The code already in SILowerI1Copies is nonsense which should be fixed
Maybe create a separate ticket?

>2. You shouldn't need to touch it
I don't see a better alternative. Do you have any suggestions?

Here is a recap of the full story. I'm listing the .ll file, the init and final DAGs, the instruction selection results and the machine code.
```diff
// .ll file
define i1 @i1_func_void() {
  %val = load i1, ptr addrspace(1) undef
  ret i1 %val
}

define void @test_call_external_i1_func_void() {
  %val = call i1 @i1_func_void()
  store volatile i1 %val, ptr addrspace(1) undef
  ret void
}

// 1. Initial DAG
Initial selection DAG: %bb.0 'test_call_external_i1_func_void:'
SelectionDAG has 17 nodes:
...
  t9: ch,glue = CALL t6, GlobalAddress:i64<ptr @i1_func_void> 0, TargetGlobalAddress:i64<ptr @i1_func_void> 0, Register:v4i32 $sgpr0_sgpr1_sgpr2_sgpr3, RegisterMask:Untyped, t6:1
  t10: ch,glue = callseq_end # D:1 t9, TargetConstant:i64<0>, TargetConstant:i64<0>, t9:1
  - t12: i1,ch,glue = CopyFromReg # D:1 t10, Register:i1 $sgpr0_sgpr1, t10:1
  t14: i64 = Constant<0>
...

+ // Note that CopyFromReg is divergent

// 2. Final DAG
Optimized legalized selection DAG: %bb.0 'test_call_external_i1_func_void:'
SelectionDAG has 21 nodes:
...
  t10: ch,glue = callseq_end # D:1 t9, TargetConstant:i64<0>, TargetConstant:i64<0>, t9:1
  - t12: i1,ch,glue = CopyFromReg # D:1 t10, Register:i1 $sgpr0_sgpr1, t10:1
      t29: i32 = zero_extend # D:1 t12
...

// 3. Instruction selection
===== Instruction selection ends:

Selected selection DAG: %bb.0 'test_call_external_i1_func_void:'
SelectionDAG has 23 nodes:
...
  t10: i1,ch,glue = ADJCALLSTACKDOWN TargetConstant:i64<0>, TargetConstant:i64<0>, t9, t9:1
  - t12: i1,ch,glue = CopyFromReg # D:1 t10:1, Register:i1 $sgpr0_sgpr1, t10:2
      t29: i32 = V_CNDMASK_B32_e64 # D:1 TargetConstant:i32<0>, TargetConstant:i32<0>, TargetConstant:i32<0>, TargetConstant:i32<1>, t12
    t27: ch = GLOBAL_STORE_BYTE<Mem:(volatile store (s8) into `ptr addrspace(1) undef`, addrspace 1)> # D:1 IMPLICIT_DEF:i64, t29, TargetConstant:i16<0>, TargetConstant:i32<0>, t12:1
...

// 4. Machine code at end of ISel
*** MachineFunction at end of ISel ***
# Machine code for function test_call_external_i1_func_void: IsSSA, TracksLiveness

...
  ADJCALLSTACKDOWN 0, 0, implicit-def dead $scc
  - %3:vreg_1 = COPY $sgpr0_sgpr1
  - %5:sreg_64_xexec = COPY %3:vreg_1
  %4:vgpr_32 = V_CNDMASK_B32_e64 0, 0, 0, 1, %5:sreg_64_xexec, implicit $exec
  %6:sreg_64 = IMPLICIT_DEF
  %7:vreg_64 = COPY %6:sreg_64
  GLOBAL_STORE_BYTE killed %7:vreg_64, killed %4:vgpr_32, 0, 0, implicit $exec :: (volatile store (s8) into `ptr addrspace(1) undef`, addrspace 1)
...
```

In `InstrEmitter::EmitCopyFromReg()` the CopyFromReg insn transformed to the following:
```
  %3:vreg_1 = COPY $sgpr0_sgpr1
  %5:sreg_64_xexec = COPY %3:vreg_1
```
If I understand you correctly, I think you want the "vreg_1" in %3 to be "sreg_64_xexec". The relevant code of `EmitCopyFromReg()` is as follows:
```
void InstrEmitter::EmitCopyFromReg(...)
  ...
  // Stick to the preferred register classes for legal types.
  if (TLI->isTypeLegal(VT))
    - UseRC = TLI->getRegClassFor(VT, Node->isDivergent());  // UseRC is set to vreg_1 here
...
          const TargetRegisterClass *RC = nullptr;
          if (i + II.getNumDefs() < II.getNumOperands()) {
            RC = TRI->getAllocatableClass(
                TII->getRegClass(II, i + II.getNumDefs(), TRI, *MF));
          }
          - if (!UseRC)  // UseRC is already set, so it's not overwritten here
            UseRC = RC;
...
  if (VRBase) {
    DstRC = MRI->getRegClass(VRBase);
  } else if (UseRC) {
    assert(TRI->isTypeLegalForClass(*UseRC, VT) &&
           "Incompatible phys register def and uses!");
    - DstRC = UseRC;   // DstRC is set to UseRC, which is vreg_1
  } else
    DstRC = SrcRC;
...
```
As you can see, because `UseRC` is set to vreg_1 near the beginning of the function, the final `DstRC` is set to vreg_1 as well.

The code for `SITargetLowering::getRegClassFor()` is as follows:
```cpp
const TargetRegisterClass *SITargetLowering::getRegClassFor(MVT VT, bool isDivergent) const {
  const TargetRegisterClass *RC = TargetLoweringBase::getRegClassFor(VT, false);
  const SIRegisterInfo *TRI = Subtarget->getRegisterInfo();
  if (RC == &AMDGPU::VReg_1RegClass && !isDivergent)  // Returns SReg_64 only for non-divergent SDNode.  
    return Subtarget->getWavefrontSize() == 64 ? &AMDGPU::SReg_64RegClass
                                               : &AMDGPU::SReg_32RegClass;
  if (!TRI->isSGPRClass(RC) && !isDivergent)
    return TRI->getEquivalentSGPRClass(RC);
  else if (TRI->isSGPRClass(RC) && isDivergent)
    return TRI->getEquivalentVGPRClass(RC);

  return RC;
}
```

I don't think changing the behavior of `SITargetLowering::getRegClassFor()` is a good idea. So, I think the current solution, i.e. skipping the instruction, `%3:sreg_64 = COPY $sgpr0_sgpr1` in SILowerI1Copies is preferred.

Pls let me know if you have any suggestions.

jwanggit86 replied:
@arsenm Any suggestions on SILowerI1Copies?

arsenm replied:
The virtual register copies are getting emitted backwards, yes.

You could hack around this in LowerFormalArguments by manually copying to a virtual register with the correct class, and then emitting a copy from that virtual register.

jwanggit86 replied:
This part is for i1 return value though, not i1 arguments.

arsenm commented on llvm/test/CodeGen/AMDGPU/z_caller2.ll:
```diff
@@ -0,0 +1,57 @@
+
+
+
+attributes #0 = { nounwind "amdgpu-no-dispatch-id" "amdgpu-no-dispatch-ptr" "amdgpu-no-implicitarg-ptr" "amdgpu-no-workgroup-id-x" "amdgpu-no-workgroup-id-y" "amdgpu-no-workgroup-id-z" "amdgpu-no-workitem-id-x" "amdgpu-no-workitem-id-y" "amdgpu-no-workitem-id-z" }
```
Don't need any of these attributes

jwanggit86 replied:
This temp test file has been removed in the latest commit.

arsenm commented on llvm/test/CodeGen/AMDGPU/z_return.ll:
```diff
@@ -0,0 +1,80 @@
+  ret void
+}
+
+attributes #0 = { nounwind }
```
Don't need this

jwanggit86 replied:
This temp test files has been removed in the latest commit.

arsenm commented on llvm/test/CodeGen/AMDGPU/z_return.ll:
```diff
@@ -0,0 +1,80 @@
+; GFX11-NEXT:    s_setpc_b64 s[30:31]
+
+  %val = call i1 @i1_func_void()
+  store volatile i1 %val, ptr addrspace(1) undef
```
Avoid undef store

jwanggit86 replied:
This temp test files has been removed in the latest commit.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -29,6 +29,39 @@
+                             .isWave64();
+
+  static const MCPhysReg I1RegList1[] = {
+      AMDGPU::SGPR0_SGPR1,   AMDGPU::SGPR2_SGPR3,   AMDGPU::SGPR4_SGPR5,
```
Do we want to only pass in even aligned registers? We could also include all the aliasing odd-based cases

jwanggit86 replied:
By "odd-based cases" do you mean something like SGPR1_SGPR2? Do you have an example where such reg allocation is used?

arsenm replied:
Yes. I mean if you had (i32 inreg %arg0, i64 inreg %arg1), you could use arg0=s0, arg1=s[1:2]

jwanggit86 replied:
This issue deals with i1 arg/return only. Inreg args are handled separately. If you are thinking about a mixed case as follows:
```
foo (i32 inreg %arg0, i1 %arg1)
```
Currently for gfx900,  %arg0 is assigned to s4, and %arg1 to s[6:7]. If in this case you want %arg1 to be given s[5:6], I suppose the list of registers can be changed from `{ ... sgpr4_sgpr5, sgpr6_sgpr7, ...}` to `{... sgpr4_sgpr5, sgpr5_sgrp6, sgpr6_sgrp7, ...}`.
Is this what you had in mind?

jwanggit86 replied:
@arsenm Pls let me know if the above is what you wanted.

arsenm replied:
Yes, the i64 is physically what is being passed, but doesn't have the alignment requirement in the argument list.

jwanggit86 replied:
It seems pairs of SGPRs that start with an odd number, e.g., SGPR5_SGPR6, are
not defined. Only pairs like SGPR4_SGPR5 are defined.

arsenm replied:
Yes, you have to unroll them into the 32-bit pieces and reassemble the virtual register

arsenm reviewed (commented):
The tests can be in one file (and possibly should just go in the existing ABI tests). They also need quite a few more tests:

1. Test i1 vectors and i1 arrays for a few different sizes
2. Test the case where the SGPR argument list will run out. In particular, test the case where a wave64 argument needs to straddle the boundary
3. Test some mixed other inreg arguments, such that the alignment requirement for the boolean uses matters. e.g. a select user will require even aligned registers, but the argument will be odd

arsenm reviewed (commented):
Also needs some AMDGPUUsage changes in the ABI sections

jwanggit86 commented:
Here is the machine code of the caller in z_return.ll at the beginning of SILowerI1Copies:
```assembly
bb.0 (%ir-block.0):
  ADJCALLSTACKUP 0, 0, implicit-def dead $scc
  %0:sreg_64 = SI_PC_ADD_REL_OFFSET target-flags(amdgpu-gotprel32-lo) @i1_func_void, target-flags(amdgpu-gotprel32-hi) @i1_func_void, implicit-def dead $scc
  %1:sreg_64_xexec = S_LOAD_DWORDX2_IMM killed %0:sreg_64, 0, 0 :: (dereferenceable invariant load (s64) from got, addrspace 4)
  %2:sgpr_128 = COPY $sgpr0_sgpr1_sgpr2_sgpr3
  $sgpr0_sgpr1_sgpr2_sgpr3 = COPY %2:sgpr_128
  SI_CALL_ISEL killed %1:sreg_64_xexec, @i1_func_void, , implicit $sgpr0_sgpr1_sgpr2_sgpr3, implicit-def $sgpr0_sgpr1
  ADJCALLSTACKDOWN 0, 0, implicit-def dead $scc
  **%3:vreg_1 = COPY $sgpr0_sgpr1**
  %5:sreg_64_xexec = COPY %3:vreg_1
  %4:vgpr_32 = V_CNDMASK_B32_e64 0, 0, 0, 1, %5:sreg_64_xexec, implicit $exec
  %6:sreg_64 = IMPLICIT_DEF
  %7:vreg_64 = COPY %6:sreg_64
  GLOBAL_STORE_BYTE killed %7:vreg_64, killed %4:vgpr_32, 0, 0, implicit $exec :: (volatile store (s8) into `ptr addrspace(1) undef`, addrspace 1)
  SI_RETURN
```
The insn starting with %3 would trigger the assert were it not for the new lines of code.

jwanggit86 committed:
Fixed (1) problems for global-isel wrt both incoming args and return
value (2) a problem in AMDCallingConv.td when no sgprs are available.

jwanggit86 commented:
The latest commit primarily fixes problems that occur when -global-isel is used. Problems with both incoming i1 args and i1 return values are fixed.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -124,7 +124,15 @@ struct AMDGPUIncomingArgHandler : public CallLowering::IncomingValueHandler {
+      unsigned CopyToBits = 32;
+
+      // When function return type is i1, it may be in a 64b register.
+      if (VA.getLocVT().getSizeInBits() == 1) {
```
getLocVT() == MVT::i1

jwanggit86 replied:
Sure.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -124,7 +124,15 @@ struct AMDGPUIncomingArgHandler : public CallLowering::IncomingValueHandler {
+
+      // When function return type is i1, it may be in a 64b register.
+      if (VA.getLocVT().getSizeInBits() == 1) {
+        if (MRI.getTargetRegisterInfo()->getRegSizeInBits(PhysReg, MRI) == 64)
```
Might as well just directly check the wave size

jwanggit86 replied:
To get wave size, in this function you need to get the MachineFunction from the MIRBuilder, then get the subtarget, and then downcast the subtarget to GCNSubtarget, and then get the wavesize. So there's also some indirections. I can make the change if you really think it's better to use wavesize.

arsenm replied:
There are more indirections going on in the physreg lookup (plus going through MRI to get TRI). It's better to just query the wavesize

jwanggit86 replied:
Ok then!

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -233,7 +241,15 @@ struct AMDGPUOutgoingArgHandler : public AMDGPUOutgoingValueHandler {
-    Register ExtReg = extendRegisterMin32(*this, ValVReg, VA);
+    Register ExtReg;
+
+    if (VA.getLocVT().getSizeInBits() == 1 &&
+        MRI.getTargetRegisterInfo()->getRegSizeInBits(PhysReg, MRI) == 64) {
```
Ditto

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -233,7 +241,15 @@ struct AMDGPUOutgoingArgHandler : public AMDGPUOutgoingValueHandler {
+
+    if (VA.getLocVT().getSizeInBits() == 1 &&
+        MRI.getTargetRegisterInfo()->getRegSizeInBits(PhysReg, MRI) == 64) {
+      ExtReg = MIRBuilder.buildAnyExt(LLT::scalar(64), ValVReg).getReg(0);
```
There shouldn't be an extension here, it should be a direct assignment

jwanggit86 replied:
OK!

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallingConv.td:
```diff
@@ -185,9 +185,12 @@ def CSR_AMDGPU_NoRegs : CalleeSavedRegs<(add)>;
 
+  CCIfType<[i1] , CCCustom<"CC_AMDGPU_Custom_I1">>,
+
+  CCIfType<[i1], CCPromoteToType<i32>>,
```
Should only promote for inreg

jwanggit86 replied:
When CC_AMDGPU_Custom_I1 fails, i.e., an i1 arg is not allocated to an SGPR. If the i1 is not promoted to i32, we end up with a COPY from a 32b VGPR to a 64b VGPR, which is rejected by SIInstrInfo::copyPhysReg():

```
renamable $vgpr0_vgpr1 = COPY killed renamable $vgpr0, implicit $exec
```

arsenm replied:
But I would still expect the first action to be covered by CCIfInReg

jwanggit86 replied:
The new CC function, CC_AMDGPU_Custom_I1, needs to have precedence over CCIfInReg. Otherwise, an "inreg i1" would be allocated to a 32b SGPR even if on a wavesize64 GPU.

arsenm replied:
which you can do? Something like CCIfType<[i1], CCIfInReg<CCPromoteToType<i32>>

jwanggit86 replied:
I think the promotion needs to be done regardless of inreg. The problem is as follows:
```
foo(a bunch of args, i1 %argX, ...)
```
If the bunch of args (i1 or inreg) before argX takes all the SGPRs, then argX is going to get a VGPR regardless of inreg or not. If it's not promoted to i32, we end up with the COPY from 32b VGPR to 64b VGPR as shown above.

arsenm replied:
No, inreg is changing the behavior of i1. The register decision is not a function of how many arguments are used.

What I want is "i1" -> SGPR pair.  "i1 inreg" -> promoted to 32-bit SGPR

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -2831,8 +2831,13 @@ SDValue SITargetLowering::LowerFormalArguments(
-    else
-      llvm_unreachable("Unexpected register class in LowerFormalArguments!");
+    else {
+      if (VT == MVT::i1 && Subtarget->isWave64())
```
You don't need to make this conditional on wave64, it's essentially checking the same thing twice

arsenm replied:
To operate within the constraints of SILowerI1Copies, we might need to find a way to get this broken into:

```
  %3:sreg_64_xexec = COPY $sgpr0_sgpr1
  %4:vreg_1 = COPY %3
  %5:sreg_64_xexec = COPY %4

```

jwanggit86 replied:
Pls see my comment on SILowerI1Copies.cpp below.

jwanggit86 committed:
Minor changes based on code review.

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -3479,6 +3484,22 @@ SDValue SITargetLowering::LowerCall(CallLoweringInfo &CLI,
+  if (!Subtarget->enableFlatScratch()) {
+    if (IsChainCallConv)
+      CCInfo.AllocateRegBlock(
+          ArrayRef<MCPhysReg>{AMDGPU::SGPR48, AMDGPU::SGPR49, AMDGPU::SGPR50,
```
AllocateReg works if you pass the register tuple value. It iterates all aliases

jwanggit86 replied:
Are you suggesting the following?
```
static const MCPhysReg RegList[] = {AMDGPU::SGPR48, AMDGPU::SGPR49, AMDGPU::SGPR50, AMDGPU::SGPR51};
CCInfo.AllocateReg(RegList);
```

jwanggit86 committed:
Additional change based on code review.

jwanggit86 committed:
Changing a vector of 4 registers to a single register.

jwanggit86 committed:
Update some test files.

jwanggit86 commented:
@arsenm Code has been updated based on your review. Pls take a look again. Some of the test files have been updated, but there are about 20 remaining.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -124,7 +124,15 @@ struct AMDGPUIncomingArgHandler : public CallLowering::IncomingValueHandler {
       // 16-bit types are reported as legal for 32-bit registers. We need to do
       // a 32-bit copy, and truncate to avoid the verifier complaining about it.
-      auto Copy = MIRBuilder.buildCopy(LLT::scalar(32), PhysReg);
+      unsigned CopyToBits = 32;
+
+      // When function return type is i1, it may be in a 64b register.
+      if (VA.getLocVT() == MVT::i1) {
+        if (MIRBuilder.getMF().getSubtarget<GCNSubtarget>().isWave64())
+          CopyToBits = 64;
+      }
+
```
This can all be in the initialization for CopyToBits

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -233,6 +241,13 @@ struct AMDGPUOutgoingArgHandler : public AMDGPUOutgoingValueHandler {
                         const CCValAssign &VA) override {
     MIB.addUse(PhysReg, RegState::Implicit);
+
+    if (VA.getLocVT() == MVT::i1 &&
```
For the i1 inreg case, do you get getLocVT == MVT::i32? that should still hit the extendRegisterMin32 path

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/irtranslator-function-args.ll:
```diff
@@ -2782,8 +2781,8 @@ define void @void_func_i1_inreg(i1 inreg %arg0) #0 {
   ; CHECK-NEXT: {{  $}}
-  ; CHECK-NEXT:   [[COPY:%[0-9]+]]:_(s32) = COPY $sgpr16
-  ; CHECK-NEXT:   [[TRUNC:%[0-9]+]]:_(s1) = G_TRUNC [[COPY]](s32)
+  ; CHECK-NEXT:   [[COPY:%[0-9]+]]:_(s64) = COPY $sgpr16_sgpr17
+  ; CHECK-NEXT:   [[TRUNC:%[0-9]+]]:_(s1) = G_TRUNC [[COPY]](s64)
```
The inreg case shouldn't change

jwanggit86 committed:
Updated calling conv such that inreg i1 is promoted to i32 before
being allocated.

jwanggit86 commented:
@arsenm In one of your comments, you mentioned promoting "inreg i1" before it's being allocated. This is done in the latest commit. Note that, on a wavefront64 GPU, an "inreg i1" argument would be allcoated to a 32b sgpr while an ordinary i1 arg would be allocated to a 64b sgpr. Is this what you had in mind?

Also I'm waiting for your further comments on SILowerI1Copies.

arsenm commented:
> @arsenm In one of your comments, you mentioned promoting "inreg i1" before it's being allocated. This is done in the latest commit. Note that, on a wavefront64 GPU, an "inreg i1" argument would be allcoated to a 32b sgpr while an ordinary i1 arg would be allocated to a 64b sgpr. Is this what you had in mind?

Yes. That is i1 inreg is not interpreted as a boolean, i1 <no inreg> is

jwanggit86 commented:
Thanks @arsenm What about SILowerI1Copies?

arsenm commented:
> Thanks @arsenm What about SILowerI1Copies?

There should be no changes there. It cannot possibly have the context necessary to do anything sensible. The code there is nonsense as it is

jwanggit86 commented:
> > Thanks @arsenm What about SILowerI1Copies?
> 
> There should be no changes there. It cannot possibly have the context necessary to do anything sensible. The code there is nonsense as it is

@arsenm So how do you want to proceed?

jwanggit86 committed:
Add an additional CopyToReg and CopyFromReg for the CopyFromReg
for the i1 return value.

jwanggit86 committed:
Revert a formatting change made by clang-format.

jwanggit86 commented:
@arsenm For i1 return value, to avoid having vreg_1 as the dest reg as in `%3:vreg_1 = COPY $sgpr0_sgpr1`, I've added a CopyToReg and another CopyFromReg for the CopyFromReg for the return val. The code now looks as follows:

Initial DAG:
```
  t10: ch,glue = callseq_end # D:1 t9, TargetConstant:i64<0>, TargetConstant:i64<0>, t9:1
  t12: i1,ch,glue = CopyFromReg # D:1 t10, Register:i1 $sgpr0_sgpr1, t10:1
  t17: i64 = Constant<0>
        t14: ch = CopyToReg # D:1 t12:1, Register:i1 %0, t12
      t15: i1,ch = CopyFromReg t14, Register:i1 %0
```

End of insn selection:
```
  t10: i1,ch,glue = ADJCALLSTACKDOWN TargetConstant:i64<0>, TargetConstant:i64<0>, t9, t9:1
  t12: i1,ch,glue = CopyFromReg # D:1 t10:1, Register:i1 $sgpr0_sgpr1, t10:2
          t14: ch = CopyToReg # D:1 t12:1, Register:i1 %0, t12
        t15: i1,ch = CopyFromReg t14, Register:i1 %0
      t32: i32 = V_CNDMASK_B32_e64 TargetConstant:i32<0>, TargetConstant:i32<0>, TargetConstant:i32<0>, TargetConstant:i32<1>, t15
```

Machine code:
```
  ADJCALLSTACKDOWN 0, 0, implicit-def dead $scc
  %4:sreg_64 = COPY $sgpr0_sgpr1
  %0:sreg_64_xexec = COPY %4:sreg_64
  %5:vgpr_32 = V_CNDMASK_B32_e64 0, 0, 0, 1, %0:sreg_64_xexec, implicit $exec
```

Pls let know your thoughts.

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -3225,6 +3230,21 @@ SDValue SITargetLowering::LowerCallResult(
+      // For i1 return value allocated to an SGPR, we want the dst reg for the
+      // above CopyFromReg not to be of VReg_1 when emitting machine code.
+      // This requires creating an addional CopyToReg followed by another
+      // CopyFromReg.
```
Maybe mention this is a workaround for the fix i1 copies

jwanggit86 replied:
Done.

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -3225,6 +3230,21 @@ SDValue SITargetLowering::LowerCallResult(
+        const SIRegisterInfo *TRI = Subtarget->getRegisterInfo();
+        MachineRegisterInfo &MRI = DAG.getMachineFunction().getRegInfo();
+
+        if (TRI->isSGPRReg(MRI, VA.getLocReg())) {
```
This should always be true if LocVT is i1

jwanggit86 replied:
So far only `CC_AMDGPU_Func/RetCC_AMDGPU_Func` have been modified such that i1 arg/return are put in SGPRs. Therefore, other calling conv functions still use VGPRs. Do you want all the others to be changed as well?

arsenm replied:
amdgpu_gfx should be the same

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -3659,6 +3679,17 @@ SDValue SITargetLowering::LowerCall(CallLoweringInfo &CLI,
     passSpecialInputs(CLI, CCInfo, *Info, RegsToPass, MemOpChains, Chain);
   }
 
+  // In code below (after call of AnalyzeCallOperands),
+  // if (!Subtarget->enableFlatScratch()), it would use either s[48:51] or
+  // s[0:3]. Therefore, before calling AnalyzeCallOperands, we may need to
+  // reserve these registers.
+  if (!Subtarget->enableFlatScratch()) {
+    if (IsChainCallConv)
+      CCInfo.AllocateReg(AMDGPU::SGPR48_SGPR49_SGPR50_SGPR51);
+    else
+      CCInfo.AllocateReg(AMDGPU::SGPR0_SGPR1_SGPR2_SGPR3);
+  }
+
```
Unrelated change?

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -15864,7 +15910,8 @@ bool SITargetLowering::isSDNodeSourceOfDivergence(const SDNode *N,
 
-    assert(Reg == FLI->DemoteRegister || isCopyFromRegOfInlineAsm(N));
+    assert(Reg == FLI->DemoteRegister || isCopyFromRegOfInlineAsm(N) ||
+           isCopyFromRegForI1Return(N));
```
This shouldn't be necessary?

jwanggit86 replied:
(1) For Debug build this is necessary. The reason is, for the i1 return value, when the additional CopyFromReg is created (line 3245 of SIISelLowering.cpp), the new virtual reg is used. Then the assert in line 15913 above would trigger without the new function `isCopyFromRegForI1Return`.
(2) Alternatively, when creating the additional CopyFromReg, we can use the physical reg from the original CopyFromReg, then the assert in line 15913 won't trigger and the new func is not needed.

arsenm reviewed (commented):
I believe we also need more tests for i1 vectors and arrays, with and without inreg / signext / zeroext

jwanggit86 commented:
> I believe we also need more tests for i1 vectors and arrays, with and without inreg / signext / zeroext

@arsenm In the following case where the func returns "zeroext i1",
```
define zeroext i1 @i1_zeroext_func_void() {
  %val = load i1, ptr addrspace(1) undef
  ret i1 %val
}
```
because of the zeroext, the return type is updated to i32 and return val allocated to a VGPR. However, on the caller side, the return val is still obtained from a SGPR. How do you want the behavior to be, using VGPR or SGPR?

cdevadas commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -124,7 +124,15 @@ struct AMDGPUIncomingArgHandler : public CallLowering::IncomingValueHandler {
+
+      // When function return type is i1, it may be in a 64b register.
+      if (VA.getLocVT() == MVT::i1) {
+        if (MIRBuilder.getMF().getSubtarget<GCNSubtarget>().isWave64())
```
Combine the two ifs?
A ternary operator for defining CopyToBits would alternately do the job.

arsenm commented:
> @arsenm In the following case where the func returns "zeroext i1",
> 
> ```
> define zeroext i1 @i1_zeroext_func_void() {
>   %val = load i1, ptr addrspace(1) undef
>   ret i1 %val
> }
> ```
> 
> because of the zeroext, the return type is updated to i32 and return val allocated to a VGPR. However, on the caller side, the return val is still obtained from a SGPR. How do you want the behavior to be, using VGPR or SGPR?

I expect this to be the boolean SGPR, and ignore the zeroext. Is inreg allowed on return values? If so, I would expect the promote-to-32-bit VGPR case

jwanggit86 committed:
This commit: (1) fixed i1 array as func return (2) fixed i1 return when
GlobalISel is used (3) zeroext/signext in i1 return is ignored (4) inreg
return of i1 is treated as i32 (5) new test files.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -88,6 +88,9 @@ struct AMDGPUOutgoingValueHandler : public CallLowering::OutgoingValueHandler {
                                         {MRI.getType(ExtReg)})
                         .addReg(ExtReg);
       ExtReg = ToSGPR.getReg(0);
+      if (VA.getLocVT() == MVT::i1 &&
+          MIRBuilder.getMF().getSubtarget<GCNSubtarget>().isWave64())
+        ExtReg = MIRBuilder.buildAnyExt(LLT::scalar(64), ExtReg).getReg(0);
```
The i1 value should have been directly assigned to the register, there shouldn't be an additional, incorrect G_ANYEXT

jwanggit86 replied:
Done.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -127,10 +130,9 @@ struct AMDGPUIncomingArgHandler : public CallLowering::IncomingValueHandler {
-        if (MIRBuilder.getMF().getSubtarget<GCNSubtarget>().isWave64())
-          CopyToBits = 64;
-      }
+      if (VA.getLocVT() == MVT::i1 &&
+          MIRBuilder.getMF().getSubtarget<GCNSubtarget>().isWave64())
+        CopyToBits = 64;
```
Ternary operator in initialization of CopyToBits

jwanggit86 replied:
Done.

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -15887,7 +15889,14 @@ static bool isCopyFromRegForI1Return(const SDNode *N) {
   if (N2->getOpcode() != ISD::CopyFromReg)
     return false;
-  SDNode *N3 = N2->getOperand(0).getNode();
+
```
This whole function shouldn't be necessary. You can't necessarily identify it's from a return context, but you shouldn't need to reverse engineer this

jwanggit86 replied:
Without this, the assert `assert(Reg == FLI->DemoteRegister || isCopyFromRegOfInlineAsm(N)` in `SITargetLowering::isSDNodeSourceOfDivergence()` would trigger in debug build. As I mentioned earlier, an alternative is to use the phys reg from the original CopyFromReg instead of a virtual reg in the new CopyFromReg.

arsenm replied:
Instead of this, you need to adjust isArgPassedInSGPR to account for the i1 case

arsenm replied:
Although this case does break the "SGPR = uniform" based logic spread through the DAG

arsenm replied:
So you also have to update isSourceOfDivergence, on top of isArgPassedinSGPR

arsenm replied:
ping here, if UniformityAnalysis was correct I don't think you need this

jwanggit86 replied:
Pls clarify. Are you saying the entire function `isCopyFromRegForI1Return()` is not needed? Also, you said "if UniformityAnalysis was correct". So it is not correct and needs fix?

arsenm replied:
Yes, this whole function should be unnecessary. isArgPassedInSGPR needs to be updated, and the use for uniformity needs to be updated to account for this non-uniform, but also in SGPR case

arsenm commented on llvm/test/CodeGen/AMDGPU/function-i1-args.ll:
```diff
@@ -0,0 +1,819 @@
+; GFX11-NEXT:    v_cmp_eq_u32_e64 s0, 1, v0
+; GFX11-NEXT:    s_waitcnt lgkmcnt(0)
+; GFX11-NEXT:    s_swappc_b64 s[30:31], s[2:3]
+  %val = load i1, ptr addrspace(1) undef
```
Don't use an undef pointer

arsenm commented on llvm/test/CodeGen/AMDGPU/function-i1-args.ll:
```diff
@@ -0,0 +1,819 @@
+  i1 %arg0, i1 %arg1, i1 %arg2, i1 %arg3, i1 %arg4, i1 %arg5, i1 %arg6, i1 %arg7,
+  i1 %arg8, i1 %arg9, i1 %arg10, i1 %arg11, i1 %arg12, i1 %arg13, i1 %arg14, i1 %arg15,
+  i1 %arg16, i1 %arg17, i1 %arg18, i1 %arg19, i1 %arg20, i1 %arg21, i1 %arg22, i1 %arg23,
+  i1 %arg24, i1 %arg25, i1 %arg26, i1 %arg27, i1 %arg28, i1 %arg29, i1 %arg30, i1 %arg31) {
```
I assume one more and the test fails?

jwanggit86 replied:
No, after all the available SGPRs are used, the rest of the args are allocated to VGPRs.

arsenm commented on llvm/test/CodeGen/AMDGPU/function-i1-args.ll:
```diff
@@ -0,0 +1,819 @@
+  ret void
+}
+
+define void @void_func_v2i1(<2 x i1> %arg0) {
```
add an inreg vector i1 test

jwanggit86 replied:
Done.

jwanggit86 committed:
This commit: (1) a fix for i1 return with GlobalISel (2) testcases.

jwanggit86 committed:
Fix formatting.

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/function-i1-args.ll:
```diff
@@ -0,0 +1,569 @@
+; GFX9-NEXT:   liveins: $sgpr4_sgpr5
+; GFX9-NEXT: {{  $}}
+; GFX9-NEXT:    [[COPY:%[0-9]+]]:_(s64) = COPY $sgpr4_sgpr5
+; GFX9-NEXT:    [[TRUNC:%[0-9]+]]:_(s1) = G_TRUNC [[COPY]](s64)
```
This should not be a truncate. We should have an s1 typed copy from sgpr4_sgpr5. Truncate is discarding bits and taking lane 0's value

jwanggit86 replied:
Without the TRUNC, the test file `GlobalISel/localizer.ll` fails with the error:
```
LLVM ERROR: unable to map instruction: %0:_(s1) = COPY $sgpr4_sgpr5 (in function: localize_internal_globals)
```
What's an "s1 typed copy"?

jwanggit86 replied:
@arsenm How do you want this to be resolved?

jwanggit86 replied:
@arsenm 
In order to allow `%0:_(s1) = COPY $sgpr4_sgpr5`, is it ok to change `AMDGPURegisterBankInfo::copyCost()` as follows, i.e., allowing the src of a COPY to be the SGPR reg bank? Note that there's no context available.
```diff
unsigned AMDGPURegisterBankInfo::copyCost(const RegisterBank &Dst,
                                           const RegisterBank &Src,
                                           TypeSize Size) const {
   // TODO: Should there be a UniformVGPRRegBank which can use readfirstlane?
   if (Dst.getID() == AMDGPU::SGPRRegBankID &&
       (isVectorRegisterBank(Src) || Src.getID() == AMDGPU::VCCRegBankID)) {
     return std::numeric_limits<unsigned>::max();
   }

   // Bool values are tricky, because the meaning is based on context. The SCC
   // and VCC banks are for the natural scalar and vector conditions produced by
   // a compare.
   //
   // Legalization doesn't know about the necessary context, so an s1 use may
   // have been a truncate from an arbitrary value, in which case a copy (lowered
   // as a compare with 0) needs to be inserted.
   if (Size == 1 &&
       (Dst.getID() == AMDGPU::SGPRRegBankID) &&
       (isVectorRegisterBank(Src) ||
-       Src.getID() == AMDGPU::SGPRRegBankID ||
        Src.getID() == AMDGPU::VCCRegBankID))
     return std::numeric_limits<unsigned>::max();

   // There is no direct copy between AGPRs.
   if (Dst.getID() == AMDGPU::AGPRRegBankID &&
       Src.getID() == AMDGPU::AGPRRegBankID)
     return 4;

   return RegisterBankInfo::copyCost(Dst, Src, Size);
 }
```

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/function-i1-args.ll:
```diff
@@ -0,0 +1,569 @@
+; GFX11-NEXT:   liveins: $sgpr0
+; GFX11-NEXT: {{  $}}
+; GFX11-NEXT:    [[COPY:%[0-9]+]]:_(s32) = COPY $sgpr0
+; GFX11-NEXT:    [[TRUNC:%[0-9]+]]:_(s1) = G_TRUNC [[COPY]](s32)
```
Same, this should have a trunc

jwanggit86 replied:
Done.

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/function-call-i1-return.ll:
```diff
@@ -0,0 +1,279 @@
+; GFX11: bb.1 (%ir-block.0):
+; GFX11-NEXT:    [[DEF:%[0-9]+]]:_(p1) = G_IMPLICIT_DEF
+; GFX11-NEXT:    [[LOAD:%[0-9]+]]:_(s1) = G_LOAD [[DEF]](p1) :: (load (s1) from `ptr addrspace(1) undef`, addrspace 1)
+; GFX11-NEXT:    [[ANYEXT:%[0-9]+]]:_(s32) = G_ANYEXT [[LOAD]](s1)
+; GFX11-NEXT:    [[INTRIN:%[0-9]+]]:_(s32) = G_INTRINSIC_CONVERGENT intrinsic(@llvm.amdgcn.readfirstlane), [[ANYEXT]](s32)
+; GFX11-NEXT:    $sgpr0 = COPY [[INTRIN]](s32)
```
This should not have any anyext, or a readfirstlane. This should be direct assignment

jwanggit86 replied:
Done.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -73,7 +79,7 @@ struct AMDGPUOutgoingValueHandler : public CallLowering::OutgoingValueHandler {
       LLT Ty = MRI.getType(ExtReg);
       LLT S32 = LLT::scalar(32);
-      if (Ty != S32) {
+      if (Ty != S32 && Ty != LLT::scalar(64)) {
```
Shouldn't need to special case the types here

jwanggit86 replied:
Done.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -63,6 +63,12 @@ struct AMDGPUOutgoingValueHandler : public CallLowering::OutgoingValueHandler {
   void assignValueToReg(Register ValVReg, Register PhysReg,
                         const CCValAssign &VA) override {
+    if (VA.getLocVT() == MVT::i1 &&
+        MIRBuilder.getMF().getSubtarget<GCNSubtarget>().isWave64()) {
```
The wave32 and wave64 paths shouldn't diverge. The special case is i1, independent of the wavesize

jwanggit86 replied:
Done.

jwanggit86 committed:
Use update_llc_test_checks.py on new test files; remove incorrect
comments in 2 new test files.

jwanggit86 committed:
For GlobalISel: (1) for incoming i1 arg/return, do not generate G_TRUNC;
however, we need to set the register class to avoid later problems with
instruction selection (2) for outgoing i1, do not differenciate between
wavesize32 and wavesize64.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -29,6 +29,38 @@
+  static bool IsWave64 =
+      State.getMachineFunction().getSubtarget<GCNSubtarget>().isWave64();
+
+  static const MCPhysReg SGPRArgsWave64[] = {
```
We should maybe unroll the 64-bit case into unaligned 32-bit pieces, but it's going to complicate things. Probably best to leave that for a follow up patch later

arsenm reviewed (commented):
This is getting closer

jwanggit86 committed:
(1) avoid using reserved ScratchRSrcReg (2) update/add testcases.

jwanggit86 committed:
Testcase updates.

jwanggit86 committed:
Fix test file after merge from main.

jwanggit86 commented on llvm/test/CodeGen/AMDGPU/GlobalISel/function-i1-args.ll:
```diff
@@ -0,0 +1,506 @@
+  store volatile i1 %arg29, ptr addrspace(1) undef
+  store volatile i1 %arg30, ptr addrspace(1) undef
+  store volatile i1 %arg31, ptr addrspace(1) undef
+
```
Add 2 more functions: `void_func_a48i1()` and `void_func_a64i1()`.

jwanggit86 commented on llvm/test/CodeGen/AMDGPU/GlobalISel/function-i1-args.ll:
```diff
@@ -0,0 +1,506 @@
+  ret void
+}
+
+define void @void_func_i1_signext(i1 signext %arg0) {
```
Added the following test functions: `void_func_zeroext_i1_i1_inreg(), void_func_i1_inreg_zeroext_i1(), void_func_signext_i1_i1_inreg(), void_func_i1_inreg_signext_i1()`.

jwanggit86 commented on llvm/test/CodeGen/AMDGPU/GlobalISel/function-call-i1-return.ll:
```diff
@@ -0,0 +1,258 @@
+; GFX9: bb.1 (%ir-block.0):
+; GFX9-NEXT:    [[DEF:%[0-9]+]]:_(p1) = G_IMPLICIT_DEF
+; GFX9-NEXT:    [[LOAD:%[0-9]+]]:_(s1) = G_LOAD [[DEF]](p1) :: (load (s1) from `ptr addrspace(1) undef`, addrspace 1)
+; GFX9-NEXT:    $sgpr0_sgpr1 = COPY [[LOAD]](s1)
```
Changed code to reserve scratchRsrcReg.

jwanggit86 commented on llvm/test/CodeGen/AMDGPU/GlobalISel/function-call-i1-return.ll:
```diff
@@ -0,0 +1,258 @@
+; GFX9-NEXT:    [[COPY:%[0-9]+]]:_(<4 x s32>) = COPY $sgpr0_sgpr1_sgpr2_sgpr3
+; GFX9-NEXT:    $sgpr0_sgpr1_sgpr2_sgpr3 = COPY [[COPY]](<4 x s32>)
+; GFX9-NEXT:    $sgpr30_sgpr31 = noconvergent G_SI_CALL [[GLOBAL]](p0), @i1_func_void, csr_amdgpu, implicit $sgpr0_sgpr1_sgpr2_sgpr3, implicit-def $sgpr0_sgpr1
+; GFX9-NEXT:    [[COPY2:%[0-9]+]]:sreg_64(s1) = COPY $sgpr0_sgpr1
```
Code has been updated to reserve scratchRSrcReg.

jwanggit86 commented on llvm/test/CodeGen/AMDGPU/GlobalISel/function-call-i1-return.ll:
```diff
@@ -0,0 +1,258 @@
+  ret void
+}
+
+define signext i1 @signext_i1_func_void() {
```
Added the following functions: `signext_inreg_i1_func_void(),  test_call_signext_inreg_i1_func_void(), zeroext_inreg_i1_func_void(),  test_call_zeroext_inreg_i1_func_void()` to both `AMDGPU/GlobalISel/function-call-i1-return.ll` and `AMDGPU/function-call-i1-return.ll`.

jwanggit86 commented on llvm/test/CodeGen/AMDGPU/GlobalISel/function-i1-args.ll:
```diff
@@ -0,0 +1,506 @@
+; GFX9-NEXT:    [[GLOBAL:%[0-9]+]]:_(p0) = G_GLOBAL_VALUE @void_func_i1  
+; GFX9-NEXT:    $sgpr0_sgpr1 = COPY [[LOAD]](s1)
+; GFX9-NEXT:    [[COPY:%[0-9]+]]:_(<4 x s32>) = COPY $sgpr0_sgpr1_sgpr2_sgpr3
+; GFX9-NEXT:    $sgpr0_sgpr1_sgpr2_sgpr3 = COPY [[COPY]](<4 x s32>)
```
Done.

jwanggit86 commented on llvm/test/CodeGen/AMDGPU/GlobalISel/function-call-i1-return.ll:
```diff
@@ -0,0 +1,258 @@
+  store volatile [2 x i1] %val, ptr addrspace(1) undef
+  ret void
+}
+
```
Added `a16i1_func_void()`. For GFX900, there are not enough SGPRs to hold a `[16 x i1]`, so the return values are allocated on the stack.

Also added `v2i1_func_void()` which returns a vector. Vectors of i1 don't use SGPRs.

jwanggit86 committed:
(1) Fix a problem with reserving ScratchRSrcD (2) update test files.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -121,6 +127,15 @@ struct AMDGPUIncomingArgHandler : public CallLowering::IncomingValueHandler {
 
+    if (VA.getLocVT() == MVT::i1) {
+      MIRBuilder.buildCopy(ValVReg, PhysReg);
+      MRI.setRegClass(ValVReg, MIRBuilder.getMF()
+                                   .getSubtarget<GCNSubtarget>()
+                                   .getRegisterInfo()
+                                   ->getBoolRC());
```
You shouldn't need to introduce a register class here, RegBankSelect should set it later

jwanggit86 replied:
Case2: i1 arg used in a branch:
After IRTranslator:
```
%0:_(s1) = COPY $sgpr4_sgpr5
%1:_(s1) = G_CONSTANT i1 true
%5:_(s32) = G_CONSTANT i32 0
%6:_(p1) = G_GLOBAL_VALUE @static.gv2
%9:_(p1) = G_GLOBAL_VALUE @static.gv0
%2:_(s1) = G_XOR %0:_, %1:_
%3:_(s1), %4:_(s64) = G_INTRINSIC_CONVERGENT_W_SIDE_EFFECTS
  intrinsic(@llvm.amdgcn.if), %2:_(s1)
G_BRCOND %3:_(s1), %bb.4
```
After RegBankSelect:
```
%0:sgpr(s1) = COPY $sgpr4_sgpr5
%20:sgpr(s32) = G_CONSTANT i32 1
%1:sgpr(s1) = G_TRUNC %20:sgpr(s32)
%21:vcc(s1) = COPY %0:sgpr(s1)
%22:vcc(s1) = COPY %1:sgpr(s1)
%2:sreg_64_xexec(s1) = G_XOR %21:vcc, %22:vcc
%4:sreg_64_xexec(s64) = SI_IF %2:sreg_64_xexec(s1), %bb.2,
  implicit-def $exec, implicit-def $scc, implicit $exec
```
After InstructionSelect:
```
%0:sreg_32 = COPY $sgpr4_sgpr5 // will cause "illegal copy"
%29:sreg_32 = S_AND_B32 1, %0:sreg_32, implicit-def
  dead $scc
%21:sreg_64_xexec = V_CMP_NE_U32_e64 0, %29:sreg_32,
  implicit $exec
%22:sreg_64_xexec = S_MOV_B64 -1
%2:sreg_64_xexec = S_XOR_B64 %21:sreg_64_xexec,
  %22:sreg_64_xexec, implicit-def dead $scc
%4:sreg_64_xexec = SI_IF %2:sreg_64_xexec, %bb.2,
  implicit-def $exec, implicit-def $scc, implicit $exec
S_BRANCH %bb.4
```

jwanggit86 replied:
Setting reg class when lowing the i1 arg works in both cases above.

arsenm replied:
The correct way to fix this is to fix RegBankSelect's mapping for COPY with s1 from SGPR physregs

jwanggit86 replied:
Modified `AMDGPURegisterBankInfo::getInstrMapping()` to set reg bank for i1 arg, instead of setting reg class in IRTranslator.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUCallLowering.cpp:
```diff
@@ -296,6 +317,35 @@ bool AMDGPUCallLowering::canLowerReturn(MachineFunction &MF,
+/// TODO: Investigate if reserving ScratchRSrcReg can be moved to calling conv
+/// functions. If so, then this function is not needed anymore -- we can just
+/// use CallLowering::determineAndHandleAssignments() as before.
+bool AMDGPUCallLowering::determineAndHandleAssignmentsLocal(
```
Instead of creating determineAndHandleAssignmentsLocal, would be better to just add an overload of determineAndHandleAssignments in the base class that takes the CCInfo as an argument. IIRC the DAG has a similar split

jwanggit86 replied:
An overloaded function with a CCState argument has been created in the base class `CallLowering`.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUISelLowering.cpp:
```diff
@@ -29,6 +29,38 @@
+      AMDGPU::SGPR28, AMDGPU::SGPR29};
+
+  assert(LocVT == MVT::i1);
+  if (unsigned Reg = IsWave64 ? State.AllocateReg(SGPRArgsWave64)
```
Could you avoid the determineAndHandleAssignments changes by using drop_front if !enableFlatScratch?

jwanggit86 replied:
If we use drop_front, doesn't that mean we are assuming that the reserved reg is s[0:3]? How about putting in here something like the following:
```
     if (!Subtarget.enableFlatScratch())
       CCInfo.AllocateReg(Info->getScratchRSrcReg());
```

arsenm replied:
Yes, could also do that

jwanggit86 replied:
Decided to create an overload function `determineAndHandleAssignments()` instead.

arsenm replied:
See above, might as well just use the 2 pieces of determineAndHandleAssignments instead

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUInstructionSelector.cpp:
```diff
@@ -131,6 +131,12 @@ bool AMDGPUInstructionSelector::selectCOPY(MachineInstr &I) const {
   Register SrcReg = Src.getReg();
 
   if (isVCC(DstReg, *MRI)) {
+    if (SrcReg.isPhysical() && SrcReg != AMDGPU::SCC) {
+      const TargetRegisterClass *DstRC = MRI->getRegClassOrNull(DstReg);
+      if (DstRC)
+        return DstRC->contains(SrcReg);
+    }
```
Can you split this into a separate change, with MIR tests? I suspect you can just directly return true.

arsenm replied:
This should be split to a separate change

arsenm commented on llvm/lib/Target/AMDGPU/SIISelLowering.cpp:
```diff
@@ -15979,6 +16022,28 @@ static bool isCopyFromRegOfInlineAsm(const SDNode *N) {
+    N3 = N3->getOperand(0).getNode();
+  } while (N3->getOpcode() == ISD::CopyFromReg);
+
+  if (N3->getOpcode() != ISD::CALLSEQ_END)
+    return false;
+  return true;
```
Return bool expression

jwanggit86 replied:
Done.

jwanggit86 committed:
Overload the function CallLowering::determineAndHandleAssignments() with
a CCState argument.

jwanggit86 committed:
For i1 arg, set reg bank in RegBankSelect, instead of setting reg class
in IRTranslator.

arsenm commented on llvm/include/llvm/CodeGen/GlobalISel/CallLowering.h:
```diff
@@ -407,6 +407,12 @@ class CallLowering {
       CallingConv::ID CallConv, bool IsVarArg,
       ArrayRef<Register> ThisReturnRegs = std::nullopt) const;
 
+  bool determineAndHandleAssignments(
```
Now that I think of it, I think it would be better to just use separate determineAssignments, and handleAssignments. determineAndHandleAssignments is already the simple case where the CCState doesn't need any custom handling

arsenm replied:
Still should undo this and just use the 2 separate pieces

jwanggit86 replied:
Done. See commit [df3a52e](https://github.com/llvm/llvm-project/pull/72461/commits/df3a52e8f182b85b80d7037b2b68c394de267fd9)

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPURegisterBankInfo.cpp:
```diff
@@ -3741,6 +3741,19 @@ AMDGPURegisterBankInfo::getInstrMapping(const MachineInstr &MI) const {
     if (!DstBank)
       DstBank = SrcBank;
 
+    // For i1 function arguments, the call of getRegBank() currently gives
+    // incorrect result. We set both src and dst banks to VCCRegBank.
+    if (!MI.getOperand(1).getReg().isVirtual() &&
+        MRI.getType(MI.getOperand(0).getReg()) == LLT::scalar(1)) {
+      DstBank = SrcBank = &AMDGPU::VCCRegBank;
+    }
+
+    // For i1 return value, the dst reg is an SReg but we need to set the reg
+    // bank to VCCRegBank.
+    if (!MI.getOperand(0).getReg().isVirtual() &&
+        SrcBank == &AMDGPU::VCCRegBank)
+      DstBank = SrcBank;
+
```
Can you also make this a separate PR (I want 2, one for the RegBankSelect and one for the InstructionSelect parts)

arsenm replied:
Ping

jwanggit86 replied:
Are you saying that the changes in `AMDGPURegisterBankInfo.cpp` should be a separate PR? Without those changes, I think some testcases will fail.

jwanggit86 committed:
For i1 function return value, an additional CopyFromReg is created to
ensure that the COPY instruction generated later on has sreg_64 as the
the destnation reg class instead of vreg_1. This commit fixes that
CopyFromReg node such that it uses the physical
register assigned to the return value instead of the new virtual
register. The purpose of doing this is to avoid having to create the
function isCopyFromRegForI1Return() that is used in
SITargetLowering::isSDNodeSourceOfDivergence() to get around the assert
therein.

arsenm reviewed (changes requested):
should split out the pieces I mentioned and needs to update UniformityAnalysis hooks

jwanggit86 committed:
Undo 6aaa564, i.e., remove the overloaded
CallLowering::determineAndHandleAssignments(). Instead, call
determineAssignments() and handleAssignments() separately.
